app/jobs/set_catkeys_and_barcodes_job.rb

```diff
@@ -0,0 +1,129 @@
+# frozen_string_literal: true
+
+##
+# Job to update/add catkey/barcodes to objects
+class SetCatkeysAndBarcodesJob < GenericJob
+  ##
+  # A job that allows a user to specify a list of pids and a list of catkeys to be associated with these pids
+  # @param [Integer] bulk_action_id GlobalID for a BulkAction object
+  # @param [Hash] params additional parameters that an Argo job may need
+  # @option params [Array] :pids required list of
+  # @option params [String] :catkeys list of catkeys to be associated 1:1 with pids in order
+  # @option params [String] :catkeys_option option to update the catkeys
+  # @option params [String] :barcodes list of barcodes to be associated 1:1 with pids in order
+  # @option params [String] :barcodes_option option to update the barcodes
+  # @option params [String] :csv_file CSV string
+  def perform(bulk_action_id, params)
+    super
+
+    update_pids, catkeys, barcodes = params_from(params)
+
+    with_bulk_action_log do |log|
+      log.puts("#{Time.current} Starting SetCatkeysAndBarcodesJob for BulkAction #{bulk_action_id}")
+      update_druid_count
+      update_pids.each_with_index do |current_druid, i|
+        change_set = ItemChangeSet.new do |change|
+          change.catkey = catkeys[i] if catkeys
+          change.barcode = barcodes[i] if barcodes
+        end
+        update_catkey_and_barcode(current_druid, change_set, log) if change_set.changed?
+      end
+      log.puts("#{Time.current} Finished SetCatkeysAndBarcodesJob for BulkAction #{bulk_action_id}")
+    end
+  end
+
+  private
+
+  def update_catkey_and_barcode(current_druid, change_set, log)
+    log.puts("#{Time.current} Beginning SetCatkeysAndBarcodesJob for #{current_druid}")
+    cocina = Dor::Services::Client.object(current_druid).find
+
+    unless ability.can?(:manage_item, cocina)
+      log.puts("#{Time.current} Not authorized for #{current_druid}")
+      return
+    end
+
+    log_update(change_set, log)
+
+    begin
+      object_client = Dor::Services::Client.object(current_druid)
+      cocina_item = object_client.find
+      state_service = StateService.new(current_druid, version: cocina_item.version)
+      open_new_version(cocina_item.externalIdentifier, cocina_item.version, version_message(change_set)) unless state_service.allows_modification?
+      ItemChangeSetPersister.update(cocina_item, change_set)
+
+      bulk_action.increment(:druid_count_success).save
+      log.puts("#{Time.current} Catkey/barcode added/updated/removed successfully")
+    rescue StandardError => e
+      log.puts("#{Time.current} Catkey/barcode failed #{e.class} #{e.message}")
+      bulk_action.increment(:druid_count_fail).save
+      nil
+    end
+  end
+
+  def dig_from_params_if_option_set(params, key)
+    params.dig(:set_catkeys_and_barcodes, key).split("\n").map(&:strip).map(&:presence) if params.dig(:set_catkeys_and_barcodes, "use_#{key}_option".to_sym) == '1'
```

Comment:
Can you say more about the case(s) `.split("\n").map(&:strip).map(&:presence)` is meant to handle? Are you trying to split multi-line input into an array of strings without whitespace padding on either side?

If so, I wonder if `params.dig(:set_catkeys_and_barcodes, key).split` might suffice:

```ruby
[1] pry(main)> " 1 \n 2 \n 3 ".split("\n").map(&:strip).map(&:presence)
=> ["1", "2", "3"]
[2] pry(main)> " 1 \n 2 \n   \n 3 ".split("\n").map(&:strip).map(&:presence)
=> ["1", "2", nil, "3"]
[3] pry(main)> " 1 \n 2 \n   \n 3 ".split
=> ["1", "2", "3"]
```

And doesn't the presence function in there mean you may have `nil`s to deal with somewhere?

Reply:
I added a comment to make clearer, but need to preserve blank lines as nils to indicate that a barcode/catkey should be removed.

Reply:
@justinlittman ah, thank you!